src/test/osd/TestPGLog.cc

```diff
@@ -1388,6 +1387,116 @@ TEST_F(PGLogTest, merge_log) {
+            | (1,4)  |  x7   |  (1,4)  |
+            |        |       |         |
+            |        |       |         |
+       head > (1,5)  |  x9   |         |
+            | MODIFY |       |         |
+            |        |       |         |
+            |        |  x9   |  (2,5)  < head
+            |        |       |  MODIFY |
+            +--------+-------+---------+
+
+      The log entry (1,5) modifies the object x9 but the olog entry (2,5) modifies
+      it and is authoritative : the log entry (1,5) is divergent.
+      This will enter Case 4.
+
+  */
+  {
+    clear();
+
+    pg_log_t olog;
+    pg_info_t oinfo;
+    pg_shard_t fromosd;
+    pg_info_t info;
+    list<hobject_t> remove_snap;
+    bool dirty_info = false;
+    bool dirty_big_info = false;
+
+    hobject_t divergent_object;
+    missing.may_include_deletes = false;
+
+    {
+      pg_log_entry_t e;
+
+      e.version = eversion_t(1, 1);
+      e.soid.set_hash(0x5);
+      log.tail = e.version;
+      log.log.push_back(e);
+      e.version = eversion_t(1, 4);
+      e.soid.set_hash(0x7);
+      log.log.push_back(e);
+      e.prior_version = eversion_t(1, 3);
+      e.version = eversion_t(1, 5);
+      e.soid.set_hash(0x9);
+      divergent_object = e.soid;
+      e.op = pg_log_entry_t::MODIFY;
+      log.log.push_back(e);
+      log.head = e.version;
+      log.index();
+
+      info.last_update = log.head;
+
+      e.version = eversion_t(1, 1);
+      e.soid.set_hash(0x5);
+      olog.tail = e.version;
+      olog.log.push_back(e);
+      e.version = eversion_t(1, 4);
+      e.soid.set_hash(0x7);
+      olog.log.push_back(e);
+      e.version = eversion_t(2, 5);
+      e.soid.set_hash(0x9);
+      e.op = pg_log_entry_t::MODIFY;
+      olog.log.push_back(e);
+      olog.head = e.version;
+    }
+
+    snapid_t purged_snap(1);
+    {
+      oinfo.last_update = olog.head;
+      oinfo.purged_snaps.insert(purged_snap);
+    }
+
+    EXPECT_FALSE(missing.have_missing());
+    EXPECT_EQ(1U, log.objects.count(divergent_object));
+    EXPECT_EQ(3U, log.log.size());
+    EXPECT_TRUE(remove_snap.empty());
+    EXPECT_EQ(log.head, info.last_update);
+    EXPECT_TRUE(info.purged_snaps.empty());
+    EXPECT_FALSE(is_dirty());
+    EXPECT_FALSE(dirty_info);
+    EXPECT_FALSE(dirty_big_info);
+
+    TestHandler h(remove_snap);
+    merge_log(oinfo, std::move(olog), fromosd, info, &h,
+              dirty_info, dirty_big_info);
+
+    /* When the divergent entry is a MODIFY and the authoritative
+       entry is a MODIFY of the same object, we can rollback all of
+       the divergent entries, then object will be added to missing.
+    */
+    EXPECT_TRUE(missing.is_missing(divergent_object));
+    EXPECT_EQ(1U, log.objects.count(divergent_object));
+    EXPECT_EQ(3U, log.log.size());
+    /* MODIFY entries from olog that are appended to the head of the log,
+       and the divergent version of the object is not removed.
+     */
+    EXPECT_TRUE(remove_snap.empty());
```

Comment:
This is the only line of this test case that fails if I revert the rest of the patch.  I actually think this condition is *wrong*.  In the local store, 0x9 is divergent and merge_log should instruct the rollbacker to remove it.  Thus, it should be in the remove_snap set.

Reply:
Actually, I think I'm wrong.  If I understand this correctly, we'll roll back the 1'5 version leaving either nothing or whatever was there previously.  missing.add_next_event in append_log_entries_update_missing will then mark it missing with the prior_version as the have field.

Reply:
If that's correct, please add a way to verify that we rolled it back (see TestHandler).

Reply:
And update the comment at the top to explain the expected behavior.